This is based on a patch taken from Dilos.

# Copyright (c) 2012-2013, Igor Kozhukhov <ikozhukhov@gmail.com>.
# All rights reserved.

diff -wpruN --no-dereference '--exclude=*.orig' a~/src/tspi/ps/tspps.c a/src/tspi/ps/tspps.c
--- a~/src/tspi/ps/tspps.c	1970-01-01 00:00:00
+++ a/src/tspi/ps/tspps.c	1970-01-01 00:00:00
@@ -48,6 +48,7 @@
 #include "spi_utils.h"
 #include "tspps.h"
 #include "tsplog.h"
+#include <libgen.h>
 
 static int user_ps_fd = -1;
 static MUTEX_DECLARE_INIT(user_ps_lock);
@@ -66,9 +67,12 @@ get_user_ps_path(char **file)
 	TSS_RESULT result;
 	char *file_name = NULL, *home_dir = NULL;
 	struct passwd *pwp;
-#if (defined (__linux) || defined (linux) || defined(__GLIBC__))
+#if (defined (__linux) || defined (linux) || defined(__GLIBC__) || defined(SOLARIS))
 	struct passwd pw;
 #endif
+#ifdef SOLARIS
+	char pwbuf[PASSWD_BUFSIZE];
+#endif
 	struct stat stat_buf;
 	char buf[PASSWD_BUFSIZE];
 	uid_t euid;
@@ -90,10 +94,22 @@ get_user_ps_path(char **file)
          * in the user's home directory, which may be shared
          * by multiple systems.
          *
-         * The directory path on Solaris is /var/tpm/userps/[EUID]/
+         * The directory path on illumos is /var/user/[USERNAME]/tpm/userps
          */
-        rc = snprintf(buf, sizeof (buf), "%s/%d", TSS_USER_PS_DIR, euid);
+
+#if (_POSIX_C_SOURCE - 0 >= 199506L) || defined(_POSIX_PTHREAD_SEMANTICS)
+	if (getpwuid_r(euid, &pw, pwbuf, sizeof (pwbuf), &pwp) != 0 ||
+	    pwp == NULL) {
+		return TSPERR(TSS_E_INTERNAL_ERROR);
+	}
 #else
+	pwp = getpwuid_r(euid, &pw, pwbuf, sizeof (pwbuf));
+	if (pwp == NULL)
+		return TSPERR(TSS_E_INTERNAL_ERROR);
+#endif
+	rc = snprintf(buf, sizeof (buf), "/var/user/%s/tpm/userps",
+	    pwp->pw_name);
+#else /* SOLARIS */
 	setpwent();
 	while (1) {
 #if (defined (__linux) || defined (linux) || defined(__GLIBC__))
@@ -138,7 +154,7 @@ get_user_ps_path(char **file)
 		if (errno == ENOENT) {
 			errno = 0;
 			/* Create the user's ps directory if it is not there. */
-			if ((rc = mkdir(buf, 0700)) == -1) {
+			if ((rc = mkdirp(buf, 0700)) == -1) {
 				LogDebugFn("USER PS: Error creating dir: %s: %s", buf,
 					   strerror(errno));
 				result = TSPERR(TSS_E_INTERNAL_ERROR);
@@ -153,8 +169,8 @@ get_user_ps_path(char **file)
 
 	/* Directory exists or has been created, return the path to the file */
 #if defined (SOLARIS)
-	rc = snprintf(buf, sizeof (buf), "%s/%d/%s", TSS_USER_PS_DIR, euid,
-		      TSS_USER_PS_FILE);
+	rc = snprintf(buf, sizeof (buf), "/var/user/%s/tpm/userps/%s",
+	    pwp->pw_name, TSS_USER_PS_FILE);
 #else
 	rc = snprintf(buf, sizeof (buf), "%s/%s/%s", home_dir, TSS_USER_PS_DIR,
 		      TSS_USER_PS_FILE);
